002-es-modules.md

```diff
@@ -0,0 +1,511 @@
+      `ModuleRecord`. Used as a means to grab modules from Node's loader/cache.
+
+## 3. Semantics
+
+### 3.1. Async loading
+
+ESM imports will be loaded asynchronously. This matches browser behavior. This means:
+
+* If a new `import` is queued up, it will never evaluate synchronously.
+* Between module evaluation within the same graph there may be other work done. Order of module evaluation within a module graph will always be preserved.
+* Multiple module graphs could be loading at the same time concurrently.
+
+### 3.2. Determining if source is an ES Module
+
+A new file type will be recognised, `.mjs`, for ES modules. This file type will be registered with IANA as an official file type, see TC39 issue. There are no known issues with browsers since they [do not determine MIME type using file extensions](https://mimesniff.spec.whatwg.org/#interpreting-the-resource-metadata).
+
+The `.mjs` file extension will not be loadable via `require()`. This means that, once the Node resolution algorithm reaches file expansion, the path for path + `.mjs` would throw an error. In order to support loading ESM in CJS files please use `import()`.
+
+### 3.3. ES Import Path Resolution
+
+
+ES `import` statements will perform non-exact searches on relative or
+absolute paths, like `require()`. This means that file extensions, and
+index files will be searched. However, ESM import specifier resolution will be done using URLs which match closer to the browser. Unlike browsers, only the `file:` protocol will be supported until network and security issues can be researched for other protocols.
+
+With `import` being URL based encoding and decoding will automatically be performed. This may affect file paths containing any of the following characters: `:`,`?`,`#`, or `%`. Details of the parsing algorithm are at the [WHATWG URL Spec](https://url.spec.whatwg.org/).
+
+* paths with `:` face multiple variations of path mutation
+* paths with `%` in their path segments would be decoded
+* paths with `?`, or `#` in their paths would face truncation of pathname
+
+All behavior differing from the [`type=module` path resolution algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#resolve-a-module-specifier) will be places in locations that would throw errors in the browser.
+
+#### 3.3.1. Algorithms
+
+##### 3.3.1.1. HostResolveImportedModule Search
+
+Notes:
+
+* The CLI has a location URL of the process working directory.
+* Paths are resolved to realpaths normally *after* all these steps.
+
+1. Apply the URL parser to `specifier`. If the result is not failure, return the result.
+2. If `specifier` does start with the character U+002F SOLIDUS (`/`), the two-character sequence U+002E FULL STOP, U+002F SOLIDUS (`./`), or the three-character sequence U+002E FULL STOP, U+002E FULL STOP, U+002F SOLIDUS (`../`)
+  1. Let `specifier` be the result of applying the URL parser to `specifier` with importing location's URL as the base URL.
```

Comment:
github's markdown renderer is confused by 2-space indent... this renders as `3.` instead of an indented `1.` 😅 

maybe 4 spaces makes it happier?